bindata/assets/alerts/client-go-requests.yaml

```
@@ -0,0 +1,23 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  name: client-go
+  namespace: openshift-kube-apiserver
+spec:
+  groups:
+    - name: control-plane-client-go
+      rules:
+        - record: client_go_request_duration:method_get:histogram_quantile99
+          expr: |
+            histogram_quantile(0.99,
+              sum(rate(
+                label_replace(rest_client_request_duration_seconds_bucket{verb="GET"},
+                  "url_host","$1$2","url","https?://([^:^/]*)(:[0-9]+)?.*?")[5m:30s]
+                )
+              ) by (url_host,job)
+            )
+        - record: client_go_request_total:network_errors:rate5m
+          expr: |
+            sum(rate(rest_client_requests_total{code="<error>"}[5m])) by (job,host)
```

Since we want insights on the proxy, I don't think we care about 4xx errors
```suggestion
            sum(rate(rest_client_requests_total{code=~"5.."}[5m])) by (job,host)
```

---

5.. means the request reached the apiserver, we have those metrics in the audit log and the apiserver metrics.
we want the ones that are not being able to contact the apiserver, these are the request that the client send and get lost

---

see
```go
// updateURLMetrics is a convenience function for pushing metrics.
// It also handles corner cases for incomplete/invalid request data.
func updateURLMetrics(ctx context.Context, req *Request, resp *http.Response, err error) {
	url := "none"
	if req.c.base != nil {
		url = req.c.base.Host
	}

	// Errors can be arbitrary strings. Unbound label cardinality is not suitable for a metric
	// system so we just report them as `<error>`.
	if err != nil {
		metrics.RequestResult.Increment(ctx, "<error>", req.verb, url)
	} else {
		//Metrics for failure codes
		metrics.RequestResult.Increment(ctx, strconv.Itoa(resp.StatusCode), req.verb, url)
	}
}
```
in https://github.com/kubernetes/kubernetes/blob/66931c9b8f11a3f223ba1890e4f390cff74ce1a6/staging/src/k8s.io/client-go/rest/request.go#L785-L799

---

wow `<error>` is so confusing, it really feels like a placeholder. Thanks for providing the details, I completely thought it was a placeholder at first

---

heh, indeed it is, let me add a comment